### PR TITLE
Fixed partial subs in Boolean expressions

### LINKED ISSUE #21971: subs is not working in logic expression
subs is not working in logic expression, for a long expression it only replacement the first match subexpression

### PR DESCRIPTION
**Pull request by oscargus**
#### References to other Issues or PRs

Fixes #21971

#### Brief description of what is fixed or changed
It is now possible to use `subs` where the expression to be replaced is part of a Boolean function.

#### Release Notes

* logic
    * `subs` works better when replacing parts of a Boolean expression.

### CODE REVIEW

File: sympy/logic/boolalg.py [hunk 1]
```diff
@@ -1136,6 +1148,14 @@ def _eval_simplify(self, **kwargs):
...
             kwargs['measure'], None)
 
+    def _eval_subs(self, old, new):
+        # If old is Xor, replace the parts of the arguments with new if all
+        # are there
+        if isinstance(old, Xor):
+            if all(a in self.args for a in old.args):
+                args = [new] + [a for a in self.args  if a not in old.args]
+                return self.func(*args)
+
```
Comment (smichr) [inline sympy/logic/boolalg.py:1158]: Might the following be better? Not sure:
```python
>>> o={1,2}
>>> s={1,2,3}
>>> sd = o.symmetric_difference(s)
>>> if not sd & old:  # all args in s are in o
...  print(sd)  # these are the args in s that are not in o
...
{3}
```

Comment (oscargus) [reply sympy/logic/boolalg.py:1158]: Probably! Felt a bit wrong to write it like that, but too tired to really think of anything else than brute force...

Comment (oscarbenjamin) [reply sympy/logic/boolalg.py:1164]: I find the logic with symmetric difference confusing. Basically this is:
```python
if old_args.issubset(self_args):
    self_args -= old_args
    self_args.add(new)
```

Comment (oscarbenjamin) [reply sympy/logic/boolalg.py:1164]: Note that `issubset` can be called with a list/tuple as argument:
```python
In [38]: {1}.issubset((1, 2, 3))
Out[38]: True

In [39]: {1}.issubset((2, 2, 3))
Out[39]: False
```
That means that the common case (condition is false) can be a fast-path that doesn't create any new sets apart from `set(old.args)`.

Comment (oscargus) [reply sympy/logic/boolalg.py:1164]: I find both set notations more confusing than the original list comprehension, but I can of course change to any of them.

Comment (smichr) [reply sympy/logic/boolalg.py:1164]: I have been confused by `symmetric_difference` but now think of it as the non-trivial difference between two sets: it gives the difference no matter what order you put the args, kind of like taking `abs(x-y)` for two scalars. 

I like @oscarbenjamin suggestion:
```
old_set = set(old)
if old_set.is_subset(args):
    args = set(args) - old_set
    args.update(new)
    return self.func(*args)
```

(with the last line unindented in Xor)

File: sympy/logic/boolalg.py [hunk 2]
```diff
@@ -1136,6 +1152,16 @@ def _eval_simplify(self, **kwargs):
         return self._apply_patternbased_simplification(rv, patterns,
             kwargs['measure'], None)
 
+    def _eval_subs(self, old, new):
+        # If old is Xor, replace the parts of the arguments with new if all
+        # are there
+        if isinstance(old, Xor):
+            old_set = set(old.args)
+            if old_set.issubset(args):
+                args = set(args) - old_set
```
Comment (smichr) [inline sympy/logic/boolalg.py:1161]:
```suggestion
            if old_set.issubset(self.args):
                args = set(self.args) - old_set
```

File: sympy/logic/boolalg.py [hunk 3]
```diff
@@ -1136,6 +1152,16 @@ def _eval_simplify(self, **kwargs):
...
+                args.update(new)
```
Comment (smichr) [inline sympy/logic/boolalg.py:1162]:
```suggestion
                args.add(new)
```

File: sympy/logic/boolalg.py [hunk 4]
```diff
@@ -872,6 +880,14 @@ def _eval_subs(self, old, new):
         if bad is not None:
             # let it raise
             bad.subs(old, new)
+        # If old is Or, replace the parts of the arguments with new if all
+        # are there
+        if isinstance(old, Or):
+            old_set = set(old.args)
+            if old_set.issubset(args):
+                args = set(args) - old_set
+                args.update(new)
```
Comment (smichr) [inline sympy/logic/boolalg.py:889]:
```suggestion
                args.add(new)
```

File: sympy/logic/boolalg.py [hunk 5]
```diff
@@ -725,6 +725,14 @@ def _eval_subs(self, old, new):
         if bad is not None:
             # let it raise
             bad.subs(old, new)
+        # If old is And, replace the parts of the arguments with new if all
+        # are there
+        if isinstance(old, And):
+            old_set = set(old.args)
+            if old_set.issubset(args):
+                args = set(args) - old_set
+                args.update(new)
```
Comment (smichr) [inline sympy/logic/boolalg.py:734]:
```suggestion
                args.add(new)
```

**Pull request merged**